src/lib/fork_id/fork_id.ml

```diff
@@ -0,0 +1,43 @@
+(* fork_id.ml -- representation of fork IDs *)
+
+open Core_kernel
+
+[%%versioned
+module Stable = struct
+  module V1 = struct
+    type t = Core_kernel.String.Stable.V1.t [@@deriving sexp]
```

Comment:
The `bin_io` serialization logic derived automatically from this versioned type is incorrect as it will not enforce the `required_length` invariant that is meant to be encapsulated by this type.

Reply:
@nholland94 it also doesn't enforce the hex string invariant

I don't want `bin_read` to raise an exception, because then attackers could bring down a node by gossipping an External_transition with a bad fork id.

I added an `is_valid` function to `Fork_id`. Can we use that in the `External_transition` validation? If so, we'd want to do it for the current fork id and the optional next fork id.

Reply:
We can add it to `External_transition` validation as a new step that must be performed and cannot be skipped. I think it may just be easier to let `bin_read` throw and exception and just provide a function from `External_transition` that wraps a `bin_read` with a try catch. So long as that "safe" function is used to read `External_transition`'s from the network, we won't need to worry about attackers crashing nodes with invalid fork ids. Serialization errors are ideally caught and handled at deserialization and not at validation, and I wouldn't want to continue shifting future serialization errors into validation checks.

Reply:
Since we're using the RPC mechanism to read external transitions, we go through `bin_read_t` of necessity.

All the response types in RPCs that include `External_transition` are also `option` types, so we can return `None` (and log something) if the fork ID is invalid.

For `New_state` gossip, we can check for a valid fork id, log the error and skip that transition if needed.

Does that sound OK?

Reply:
Sounds fine for now. We will likely need to get more control over how the RPC library handles serialization errors from bin_io, but this sounds like a good way to move forward for now.

Reply:
I've started on this, but it's feeling like the wrong separation of concerns. We're doing data validation of the fork id in what's essentially networking code that otherwise does no validation, and in several places.

We should be to do such validation in just one place.

Reply:
Actually, maybe I can do the #4200 stuff in these places, which I'd feel more comfortable with. We could check that the fork ID matches the current fork ID, and sure, check the validity at the same time. I'm OK with that.